File: app/src/main/java/com/example/tictactoe/MainActivity.java
package com.example.tictactoe;

import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.GridLayout;
import android.widget.TextView;

import androidx.appcompat.app.AppCompatActivity;

public class MainActivity extends AppCompatActivity {

    private String[][] board = new String[3][3]; // Represents the game board
    private boolean isPlayerX = true; // X starts first
    private boolean gameActive = true;

    private TextView gameStatus;
    private Button playAgainButton;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        gameStatus = findViewById(R.id.gameStatus);
        playAgainButton = findViewById(R.id.playAgainButton);
        GridLayout gameGrid = findViewById(R.id.gameGrid);

        // Initialize grid buttons
        for (int i = 0; i < gameGrid.getChildCount(); i++) {
            final int row = i / 3;
            final int col = i % 3;
            Button cell = (Button) gameGrid.getChildAt(i);
            cell.setOnClickListener(view -> makeMove(cell, row, col));
        }

        // Reset the game
        playAgainButton.setOnClickListener(view -> resetGame(gameGrid));
    }

    private void makeMove(Button cell, int row, int col) {
        if (!gameActive || board[row][col] != null) return; // Ignore invalid moves

        // Update the board and UI
        board[row][col] = isPlayerX ? "X" : "O";
        cell.setText(board[row][col]);

        // Check for winner
        if (checkWinner()) {
            gameStatus.setText((isPlayerX ? "X" : "O") + " wins!");
            gameActive = false;
            playAgainButton.setVisibility(View.VISIBLE);
        } else if (isBoardFull()) {
            gameStatus.setText("It's a draw!");
            gameActive = false;
            playAgainButton.setVisibility(View.VISIBLE);
        } else {
            isPlayerX = !isPlayerX; // Switch turn
            gameStatus.setText((isPlayerX ? "X" : "O") + " Play");
        }
    }

    private boolean checkWinner() {
        // Check rows and columns
        for (int i = 0; i < 3; i++) {
            if (board[i][0] != null && board[i][0].equals(board[i][1]) && board[i][1].equals(board[i][2])) return true;
            if (board[0][i] != null && board[0][i].equals(board[1][i]) && board[1][i].equals(board[2][i])) return true;
        }

        // Check diagonals
        if (board[0][0] != null && board[0][0].equals(board[1][1]) && board[1][1].equals(board[2][2])) return true;
        if (board[0][2] != null && board[0][2].equals(board[1][1]) && board[1][1].equals(board[2][0])) return true;

        return false;
    }

    private boolean isBoardFull() {
        for (String[] row : board) {
            for (String cell : row) {
                if (cell == null) return false;
            }
        }
        return true;
    }

    private void resetGame(GridLayout gameGrid) {
        // Clear the board
        for (int i = 0; i < gameGrid.getChildCount(); i++) {
            Button cell = (Button) gameGrid.getChildAt(i);
            cell.setText("");
            cell.setEnabled(true);
        }
        board = new String[3][3];
        isPlayerX = true;
        gameActive = true;
        gameStatus.setText("X Play");
        playAgainButton.setVisibility(View.GONE);
    }
}

File: app/src/main/java/com/example/tictactoe/MainActivity.kt
package com.example.tictactoe

import android.os.Bundle
import android.view.View
import android.widget.Button
import android.widget.GridLayout
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity

class MainActivity : AppCompatActivity() {
    private val board = Array(3) { arrayOfNulls<String>(3) }
    private var isPlayerX = true
    private var gameActive = true

    private lateinit var gameStatus: TextView
    private lateinit var playAgainButton: Button
    private lateinit var gameGrid: GridLayout

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        gameStatus = findViewById(R.id.gameStatus)
        playAgainButton = findViewById(R.id.playAgainButton)
        gameGrid = findViewById(R.id.gameGrid)

        for (i in 0 until gameGrid.childCount) {
            val row = i / 3
            val col = i % 3
            val cell = gameGrid.getChildAt(i) as Button
            cell.setOnClickListener {
                makeMove(cell, row, col)
            }
        }

        playAgainButton.setOnClickListener {
            resetGame()
        }
    }

    private fun makeMove(cell: Button, row: Int, col: Int) {
        if (!gameActive || board[row][col] != null) return

        board[row][col] = if (isPlayerX) "X" else "O"
        cell.text = board[row][col]

        when {
            checkWinner() -> {
                gameStatus.text = "${if (isPlayerX) "X" else "O"} wins!"
                gameActive = false
                playAgainButton.visibility = View.VISIBLE
            }
            isBoardFull() -> {
                gameStatus.text = "It's a draw!"
                gameActive = false
                playAgainButton.visibility = View.VISIBLE
            }
            else -> {
                isPlayerX = !isPlayerX // Switch turn
                gameStatus.text = "${if (isPlayerX) "X" else "O"} Play"
            }
        }
    }

    private fun checkWinner(): Boolean {
        for (i in 0..2) {
            if (board[i][0] != null && board[i][0] == board[i][1] && board[i][1] == board[i][2]) return true
            if (board[0][i] != null && board[0][i] == board[1][i] && board[1][i] == board[2][i]) return true
        }

        if (board[0][0] != null && board[0][0] == board[1][1] && board[1][1] == board[2][2]) return true
        if (board[0][2] != null && board[0][2] == board[1][1] && board[1][1] == board[2][0]) return true

        return false
    }

    private fun isBoardFull(): Boolean {
        for (row in board) {
            for (cell in row) {
                if (cell == null) return false
            }
        }
        return true
    }

    private fun resetGame() {
        for (i in 0 until gameGrid.childCount) {
            val cell = gameGrid.getChildAt(i) as Button
            cell.text = ""
        }

        for (i in board.indices) {
            for (j in board[i].indices) {
                board[i][j] = null
            }
        }

        isPlayerX = true
        gameActive = true
        gameStatus.text = "X Play"
        playAgainButton.visibility = View.GONE
    }
}
